### ISSUE TITLE
can you please explain a bit of `gen_flow` ?

### ISSUE THREAD

**Issue by moonman925** [MokkeMeguru/glow-realnvp-tutorial#2]
I changed parameters in `gen_flow` like `gen_flow([32,32,3])`, and it threw an error saying shape (40,) and (24,) are incompatible. I want to use your model to train on 3 channels images and it failed.

**Comment by MokkeMeguru** [comment]
Thanks your advice!

I failed the calculation about input shapes' in below function.
```
    def _gen_input_shapes(input_shape, level):
        input_shape = input_shape
        input_shapes = []
        for i in range(level):
            input_shape = [
                input_shape[0] // 2,
                input_shape[1] // 2,
                input_shape[2] * 4 - i * 8,
            ]
            input_shapes.append(input_shape)
        print('input_shapes', input_shapes)
        return input_shapes
``` 

invalid
```
input_shapes [[16, 16, 12], [8, 8, 40], [4, 4, 144]]
```

⇒ correct
```
input_shapes [[16, 16, 12], [8, 8, 24], [4, 4, 48]]
```

**Comment by MokkeMeguru** [comment]
You can see 
[My Multi-Scale Architecture's image](https://raw.githubusercontent.com/MokkeMeguru/glow-realnvp-tutorial/89344d6fb7dbdf5c379fbee97cf14c28f926f9a0/examples/img/multi-scale-arch.png)

So Input shape will be

[x, y, c]
=> [1/2 x , 1/2 y, 4 c] (Squeeze)
=> [x' = 1/2 x, y' = 1/2 y, z' = 2 c] (Blockwise)
=> [1/2 x', 1/2 y', 4z'] (Squeeze)
=> ...

**Comment by MokkeMeguru** [comment]
Update by https://github.com/MokkeMeguru/glow-realnvp-tutorial/commit/9580a9f47bf9f4a81cec51791a3974aaaa22a958

Can you check my update?

**Comment by moonman925** [comment]
> Update by [9580a9f](https://github.com/MokkeMeguru/glow-realnvp-tutorial/commit/9580a9f47bf9f4a81cec51791a3974aaaa22a958)
> 
> Can you check my update?

There was something wrong here 
```
def _gen_input_shapes(input_shape, level):
        input_shape = input_shape
        input_shapes = []
        for i in range(level):
            input_shape = [
                input_shape[0] // 2,
                input_shape[1] // 2,
                input_shape[2] * 2,
            ]
            input_shapes.append(input_shape)
        return input_shapes

    input_shapes[-1] = input_shapes[-1] * 2 # Here input_shapes is not defined yet , I believe it should be input_shape without an 's'
    input_shapes = _gen_input_shapes(input_shape, level)
```

You did a really great tutorial! THX!

**Comment by moonman925** [comment]
I wonder how to save the model. Cuz the `flow` is a distribution not a keras model or something else, you know how to save the trained model?

**Comment by MokkeMeguru** [comment]
Simply, you can use tf.train.checkpoint

Ok, I will write a cheep code for you now.
But I say 'YOU SHOULD USE PYTORCH (pyxyz)' Because Tensorflow probability's functions are TOO BAD.
For example, if you write a new flow model for a research paper,  you cannot compare log probability with previous papers' value (you should re-implement these model) sinse tfd.TransformDistribution's log_prob function is something wrong. ref. https://github.com/tensorflow/probability/issues/576

**Comment by MokkeMeguru** [comment]
Here is the example
https://github.com/MokkeMeguru/glow-realnvp-tutorial/blob/master/realworld/train.py#L94-L106

**Comment by MokkeMeguru** [comment]
```
def _gen_input_shapes(input_shape, level):
        input_shape = input_shape
        input_shapes = []
        for i in range(level):
            input_shape = [
                input_shape[0] // 2,
                input_shape[1] // 2,
                input_shape[2] * 2,
            ]
            input_shapes.append(input_shape)
        return input_shapes

    input_shape[-1] = input_shape[-1] * 2  # input_shapes[-1] = input_shapes[-1] * 2 
    input_shapes = _gen_input_shapes(input_shape, level)
```
and fixed bug... thanks

**Comment by moonman925** [comment]
> Simply, you can use tf.train.checkpoint
> 
> Ok, I will write a cheep code for you now.
> But I say 'YOU SHOULD USE PYTORCH (pyxyz)' Because Tensorflow probability's functions are TOO BAD.
> For example, if you write a new flow model for a research paper, you cannot compare its log probability with previous papers' value (you should re-implement these models) since tfd.TransformDistribution's log_prob function is something wrong. ref. [tensorflow/probability#576](https://github.com/tensorflow/probability/issues/576)


Thank you so much for all these advices! I'll consider using pytorch in the future!

**Comment by moonman925** [comment]
Hey! There is one more thing I want to consult you. Is that the way you calculate the loss same as the way the author of glow did?  Is the target distribution (learned one) about one image or all training images? I mean, the target distribution is a distribution for one exact image of all training images? I do really hope that you can help me out with this... sorry to  trouble you this way tho. @MokkeMeguru

**Comment by MokkeMeguru** [comment]
OK, let's write formula.

First, Log_probability definition


log p_X (X=x) = log p_Z (Z=f(x)) + log |det df(x) / d(x)|  

- x means a sample image  H x W x C

-------------------------------------------------------------------------------------------------------------
Below is a implementation note.

# First
log p_Z(Z=f(x)) => tf.reduce_sum(tfd.Normal(0, 1).log_prob(z), axis=[1, 2, 3]) 
or
log p_Z(Z=f(x)) => tfd.MultivariateNormalDiag(tf.zeros([H*W*C]), tf.ones([H*W*C])).log_prob(z)

- shape is  [batch_size]

# Second
and log |def df(x) / d(x)| 

flow.log_det_jacobian(x) => something wrong because Tensorflow Probability

- **shape should be [batch_size]**

# Third
and ALSO (we often forget, and I forgot) WE SHOULD ADD PREPROCESS's LOG_DET_JACOBIAN
https://github.com/openai/glow/blob/master/model.py#L172

formula: [0, 255] -> [-0.5, 0.5]

implementation: - log(256.) * H * W * C 

- shape is [batch-size] (broadcast)

# Fourth
So negative log likelihood loss will be 

nll = - (log p_Z + log_det_jacobian + preprocess_log_det_jacobian) 

- shape is [batch-size]

So test-loss (paper loss) will calculated as follows

mean(nll)

-------------------------------------------------------------------
My implementation misstake
1. forgot Third Loss (but this is a constant value is no related with training)
2. TFP's Strange behavior

**Comment by MokkeMeguru** [comment]
And Also, you also can read this issue
https://github.com/openai/glow/issues/43 (this issue explained the third loss)

**Comment by MokkeMeguru** [comment]
And if you want you use bits/dims instead of NLL, you should read
https://github.com/rosinality/glow-pytorch/issues/3

**Comment by moonman925** [comment]
Nice! The final distribution after training represents the whole training images' distribution right?

**Comment by moonman925** [comment]
@MokkeMeguru  Hey! Sorry to bother you again. I found you added ffjord jupyter notebook. I read the original paper but still have no clue how to implement it to images training. 

Do you know the constraint of ODEnet ('MLP_ODE' in that notebook). I tried implementing Conv2d instead of dense but it throw shape incompare error.
 
I stuck for a while and have no tutorial on such kinda area. So.. Like I said, I am sorry to bother you but hopefully you can share me with some thoughts.

**Comment by MokkeMeguru** [comment]
Sorry. I didn't find your post. I now try it with some official codes for Neural ODE. 
Can you Please Open New Issue of Notification for me?

**Comment by MokkeMeguru** [comment]
And If you have more question, You can open the issue at https://github.com/MokkeMeguru/TFGENZOO. 
This repository has some readable implementation of normalizing flow (I think...)
